tests/models/test_pyro.py

```diff
@@ -62,20 +48,64 @@ def __init__(self, in_features, out_features):
             .to_event(1)
         )
 
-    def forward(self, x, y):
-        sigma = pyro.sample("sigma", dist.Uniform(self.zero, self.ten))
+    def create_plates(self, x, y, ind_x):
+        """Function for creating plates is needed when using AutoGuides
+        and should have the same call signature as model"""
+        return pyro.plate("obs_plate", size=self.n_obs, dim=-2, subsample=ind_x)
+
+    def list_obs_plate_vars(self):
+        """Create a dictionary with:
+        1. "name" - the name of observation/minibatch plate;
+        2. "in" - indexes of model args to provide to encoder network when using amortised inference;
+        3. "sites" - dictionary with
+            keys - names of variables that belong to the observation plate (used to recognise
+             and merge posterior samples for minibatch variables)
+            values - the dimensions in non-plate axis of each variable (used to construct output
+             layer of encoder network when using amortised inference)
+        """
+
+        return {
+            "name": "obs_plate",
+            "in": [0],  # model args index for expression data
+            "sites": {"per_cell_weights": 1},
+        }
+
+    @staticmethod
+    def _get_fn_args_from_batch(tensor_dict):
+        x = tensor_dict[_CONSTANTS.X_KEY]
+        y = tensor_dict[_CONSTANTS.LABELS_KEY]
+        ind_x = tensor_dict["ind_x"].long().squeeze()
+        return (x, y, ind_x), {}
+
+    def forward(self, x, y, ind_x):
+
+        obs_plate = self.create_plates(x, y, ind_x)
+
+        sigma = pyro.sample("sigma", dist.Exponential(self.one))
+
         mean = self.linear(x).squeeze(-1)
-        with pyro.plate("data", size=self.n_obs, subsample_size=x.shape[0]):
+
+        if self.per_cell_weight:
+            with obs_plate:
+                per_cell_weights = pyro.sample(
+                    "per_cell_weights", dist.Normal(self.zero, self.one)
+                )
+                mean = mean + per_cell_weights.squeeze(-1)
+
+        with obs_plate:
             pyro.sample("obs", dist.Normal(mean, sigma), obs=y)
         return mean
 
 
 class BayesianRegressionModule(PyroBaseModuleClass):
-    def __init__(self, in_features, out_features):
+    def __init__(self, **kwargs):
 
         super().__init__()
-        self._model = BayesianRegressionPyroModel(in_features, out_features)
-        self._guide = AutoDiagonalNormal(self.model)
+        self._model = BayesianRegressionPyroModel(**kwargs)
+        self._guide = AutoNormal(
+            self.model, init_loc_fn=init_to_mean, create_plates=self.model.create_plates
+        )
+        self._get_fn_args_from_batch = self._model._get_fn_args_from_batch
```

Comment:
going back to previous point... why is this necessary?, why not define the static method here instead?

Reply:
We discussed this on slack - because it is easier to keep model-specific code in pyro model class - that way scVI module class can be reused for multiple similar models.